odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
+                p = 0.0
+            poffsets_sky.append(p)
+
+            if 'telescope:q' in step_keys:
+                q = float(step['telescope:q'])
+            else:
+                q = 0.0
+            qoffsets_sky.append(q)
+
+        if 'observe:coadds' in step_keys:
+            ncoadds = int(step['observe:coadds'])
+        else:
+            ncoadds = 1
+        poffsets.append(p)
+        qoffsets.append(q)
+
+        exptimes.append(float(step['observe:exposureTime']))
+        coadds.append(ncoadds)
+
+    mode = obsmode(config)
+    if verbose:
+        print(config)
+        print(mode)
+
+    if config['inst'] == 'GPI':
+        do_not_split = True
+
+    # Analyze sky offset patterns using auto-correlation
+    # The lag is the length of any pattern, 0 means no repeating pattern
+    plag = 0
+    qlag = 0
+    offset_lag = 0
+    if do_not_split:
+        offset_lag = nsteps
+    else:
+        if len(poffsets_sky) > 1:
+            plag = autocorr_lag(np.array(poffsets_sky))
+        if len(qoffsets_sky) > 1:
+            qlag = autocorr_lag(np.array(qoffsets_sky))
+
+        # Special cases
+        if plag == 0 and qlag == 0 and len(qoffsets_sky) == 4:
+            # single ABBA pattern, which the auto-correlation won't find
+            if qoffsets_sky[0] == qoffsets_sky[3] and qoffsets_sky[1] == qoffsets_sky[2]:
+                qlag = 4
+        elif len(qoffsets_sky) == 2:
+            # If only two steps, put them together, might be AB, also silly to split only two steps
+            qlag = 2
+
+        offset_lag = qlag
+        if plag > 0 and plag != qlag:
+            offset_lag = 0
+
+        # Turn off offset groups if the length is longer than some limit
+    #     if offset_lag > 0 and plag > 0 and qlag > 0 and \
+    #        np.sum(np.asarray(exptimes[0:offset_lag]) * np.asarray(coadds[0:offset_lag])) > 600.:
+    #         offset_lag = 0
+
+    if verbose:
+        print('Offset lags: ', plag, qlag, offset_lag, file=fid)
+
+    # Group by changes in exptimes/coadds?
+    exptime_groups = False
+    #     if len(uniquelist(exptimes)) > 1 or len(uniquelist(coadds)) > 1:
+    #         exptime_groups = True
+
+    # Second pass to determine atom properties
+    # Write results to the Excel worksheet if given
+    if ws is not None:
+        # Columns
+        columns = ['datalab', 'class', 'type', 'inst', 'exec_time', 'exptime', 'coadds', 'fpu', 'filter',
+                   'disperser', 'wavelength', 'p', 'q', 'guiding', 'qa_state', 'atom']
+        row = 1
+        for jj, col in enumerate(columns):
+            _ = ws.cell(column=jj + 1, row=row, value="{0}".format(col))
+        row += 1
+
+    npattern = offset_lag
+    noffsets = 0
     for ii, step in enumerate(sequence):
         nextatom = False
-        qoffsets = False
 
         datalab = step['observe:dataLabel']
         if datalab in datalabels:
             qastate = obslog[datalab]['qaState']
         else:
             qastate = 'NONE'
```

Comment:
Through here and below, we have a `QAState` enum in the mini-model: I think we should be using these values?